### DISCUSSION TITLE
Add method to get transaction result (return value or revert reason)

### DISCUSSION THREAD

**Discussion by frolic** [wevm/viem#462] [Idea / Feature Request]
For a given transaction hash, I'd like to retrieve the result of the transaction (return value if success, error string or error data if revert). Doing this myself feels a little boilerplate-y, especially juggling the differences between viem and RPC (e.g. `from` vs `account` vs `sender`, `to` vs `address`), and the code looks an awful lot like a viem's other methods.

```ts
async function getTransactionResult(hash: Hex) {
  const tx = await getTransaction(publicClient, hash);
  const txReceipt = await getTransactionReceipt(publicClient, hash);
  const { functionName, args } = decodeFunctionData({ abi, data: tx.input });
  try {
    return publicClient.simulateContract({
      account: tx.from,
      address: tx.to!,
      abi,
      functionName,
      args,
      // value: tx.value, // TS doesn't like me to including this
      blockNumber: receipt.blockNumber,
      // do we need to include nonce, gas price, etc. to properly simulate?
    });
  } catch (error) {
    throw getContractError(error as BaseError, {
      abi,
      address: tx.to!,
      sender: tx.from,
      functionName,
      args,
    });
  }
}
```

Side note that it feels very weird to me that `simulateContract` is our best option here for retrieving the transaction result. A tx at the beginning of the block might have a different result than at the end, or a tx that reverted at the end may not revert at the beginning, so I assume `simulateContract` isn't entirely accurate in some cases.

Aside from tracing (which seems to be maybe not widely supported by RPCs or behind a specific payment tier), is there really no other way to determine the outcome of a transaction?

**Comment by tmm** [comment]
Solid write up! Just a quick question for now: What's the value for `abi`?

> // value: tx.value, // TS doesn't like me to including this

**Reply by frolic** [reply]
Oops, sorry I missed this question. I think what you're hinting at is the `abi` doesn't include a method that takes a `value`, so it's erroring. That makes sense!

**Comment by norswap** [comment]
Adding a +1 here, this would be useful.

My understanding of the problem is that the *standard* JSON-RPC interface does not provide a way to get the return value or the revert code from a `eth_sendRawTransaction` call.

It's possible to retrieve the revert data by parsing the logs, but this takes an additional JSON-RPC query (unless it's possible to set it up as some kind of subscription? I'm out of my depth here). As for the return value I'm not even sure that it's possible.
